client/credentials.go

```diff
@@ -0,0 +1,108 @@
+package client
+
+import (
+	"crypto/sha256"
+	"encoding/json"
+	"fmt"
+	"hash"
+	"net/http"
+	"net/url"
+	"strings"
+
+	"github.com/tent/hawk-go"
+)
+
+// Credentials for taskcluster and methods to sign requests.
+type Credentials struct {
+	ClientID         string   `json:"clientId"`
+	AccessToken      string   `json:"accessToken"`
+	Certificate      string   `json:"certificate"`
+	AuthorizedScopes []string `json:"authorizedScopes"`
+}
+
+// PayloadHash creates payload hash calculator for given content-type
+func PayloadHash(contentType string) hash.Hash {
+	a := hawk.Auth{
+		Credentials: hawk.Credentials{
+			Hash: sha256.New,
+		},
+	}
+	return a.PayloadHash(contentType)
+}
+
+type certificate struct {
+	Version   int      `json:"version"`
+	Scopes    []string `json:"scopes"`
+	Start     int64    `json:"start"`
+	Expiry    int64    `json:"expiry"`
+	Seed      string   `json:"seed"`
+	Signature string   `json:"signature"`
+	Issuer    string   `json:"issuer,omitempty"`
+}
+
+type ext struct {
+	Certificate      *certificate `json:"certificate,omitempty"`
+	AuthorizedScopes *[]string    `json:"authorizedScopes,omitempty"`
+}
+
+func (c *Credentials) newAuth(method, url string, h hash.Hash) (*hawk.Auth, error) {
+	// Create a hawk auth
+	a, err := hawk.NewURLAuth(url, &hawk.Credentials{
+		ID:   c.ClientID,
+		Key:  c.AccessToken,
+		Hash: sha256.New,
+	}, 0)
+	if err != nil {
+		return nil, err
+	}
+	a.Method = method
```

Comment:
You also need to set a `nonce`...

```
    b := make([]byte, 8)
    _, err = io.ReadFull(rand.Reader, b)
    if err != nil {
        panic(err)
    }   
    a.Nonce = base64.StdEncoding.EncodeToString(b)[:8]
```

I even tested to be sure that `newAuth` isn't setting one, and confirmed it isn't:

```
package main

import (
        "fmt"
        "net/http"
        "net/http/httputil"
        "os"

        "github.com/taskcluster/taskcluster-cli/client"
)

func main() {
        c := &client.Credentials{
                ClientID:    os.Getenv("TASKCLUSTER_CLIENT_ID"),
                AccessToken: os.Getenv("TASKCLUSTER_ACCESS_TOKEN"),
                Certificate: os.Getenv("TASKCLUSTER_CERTIFICATE"),
        }
        req, _ := http.NewRequest("GET", "https://secrets.taskcluster.net/v1/secrets", nil)
        c.SignRequest(req, nil)
        req.Header.Write(os.Stdout)
        resp, _ := (&http.Client{}).Do(req)
        data, _ := httputil.DumpResponse(resp, true)
        fmt.Println(string(data))
}
```

gives me:

```
Authorization: Hawk id="mozilla-ldap/pmoore@mozilla.com/dev", mac="************", ts="1470333769", nonce=""
HTTP/1.1 200 OK
Content-Length: 19
Access-Control-Allow-Headers: X-Requested-With,Content-Type,Authorization,Accept,Origin
Access-Control-Allow-Methods: OPTIONS,GET,HEAD,POST,PUT,DELETE,TRACE,CONNECT
Access-Control-Allow-Origin: *
Access-Control-Request-Method: *
Connection: keep-alive
Content-Type: application/json; charset=utf-8
Date: Thu, 04 Aug 2016 18:02:50 GMT
Etag: W/"13-575462c4"
Server: Cowboy
Via: 1.1 vegur
X-Powered-By: Express

{
  "secrets": []
}
```

(see - no `nonce` and no secrets)

However, if I add my patch above, I get the correct output:

```
Authorization: Hawk id="mozilla-ldap/pmoore@mozilla.com/dev", mac="*******", ts="1470333645", nonce="2Bd7/MZ4"
HTTP/1.1 200 OK
Content-Length: 1468
Access-Control-Allow-Headers: X-Requested-With,Content-Type,Authorization,Accept,Origin
Access-Control-Allow-Methods: OPTIONS,GET,HEAD,POST,PUT,DELETE,TRACE,CONNECT
Access-Control-Allow-Origin: *
Access-Control-Request-Method: *
Connection: keep-alive
Content-Type: application/json; charset=utf-8
Date: Thu, 04 Aug 2016 18:00:47 GMT
Etag: W/"oV4NVLg3rFx8j5I+ZVD99g=="
Server: Cowboy
Via: 1.1 vegur
X-Powered-By: Express

{
  "secrets": [
    "garbage/pmoore/.......",
    "project/releng/......",
    "project/releng/......",
    ....
    "project/taskcluster/.......",
    "project/taskcluster/.......",
    ....
  ]
}
```

so it looks like server side we require the nonce, so ... you'll need to add it. ;)